kernel/uid_gid.c

```diff
@@ -0,0 +1,775 @@
+    thread->egid = 0;
+
+    // root succeeds
+    if (gid == 0)
+    {
+        ret = 1;
+        goto done;
+    }
+
+    /* errors reading file return -1 */
+    ret = -1;
+
+    /* get file length */
+    if (myst_syscall_stat("/etc/passwd", &file_stat) != 0)
+        goto done;
+
+    file_length = file_stat.st_size;
+
+    buffer = malloc(file_length + 1);
+    if (buffer == NULL)
+        goto done;
+    buffer[file_length] = '\0';
+
+    fd = myst_syscall_open("/etc/passwd", O_RDONLY, 0);
+    if (fd == -1)
+        goto done;
+
+    /* read into buffer */
+    if (myst_syscall_read(fd, buffer, file_length) != file_length)
+        goto done;
+
+    /* Failures from now are not found errors */
+    ret = 0;
+
+    char* start_line = buffer;
+    while (start_line)
+    {
+        char* group_name = start_line;
+        char* password = strchr(group_name, ':');
+        if ((password == NULL) || (*password != ':'))
+        {
+            goto done;
+        }
+        *password = '\0';
+        password++;
+        char* gid_str = strchr(password, ':');
+        if ((gid_str == NULL) || (*gid_str != ':'))
+        {
+            goto done;
+        }
+        *gid_str = '\0';
+        gid_str++;
+        char* list_users = strchr(gid_str, ':');
+        if ((list_users == NULL) || (*list_users != ':'))
+        {
+            goto done;
+        }
+        *list_users = '\0';
+        list_users++;
+        start_line = strchr(list_users, '\n');
+
+        // need to handle end of line and end of file properly */
+        if ((start_line != NULL) && (*start_line == '\n'))
+        {
+            *start_line = '\0';
+            start_line++;
+        }
+        else
+        {
+            /* we must be at end of file */
+            start_line = NULL;
+        }
+
+        if (gid == (uid_t)atoi(gid_str))
+        {
+            ret = 1;
+            goto done;
+        }
+    }
+
+done:
+    if (fd != -1)
+        myst_syscall_close(fd);
+
+    thread->egid = save_egid;
+    thread->euid = save_euid;
+
+    if (buffer)
+    {
+        memset(buffer, 0, file_length);
+        free(buffer);
+    }
+    return ret;
+}
+
+long myst_syscall_getuid()
+{
+    myst_thread_t* thread = myst_thread_self();
+    return thread->uid;
+}
+
+long myst_syscall_setuid(uid_t uid)
+{
+    long ret = 0;
+    myst_thread_t* thread = myst_thread_self();
+
+    /* EINVAL if not valid uid */
+    if (myst_valid_uid(uid) <= 0)
+    {
+        ret = -EINVAL;
+    }
+    /* handle if euid is root (TODO: or CAP_SETUID is set), set all UIDs
+     * to specified uid */
+    else if (thread->euid == 0)
+    {
+        thread->uid = uid;
+        thread->euid = uid;
+        thread->savuid = uid;
+        thread->fsuid = uid;
+    }
+    /* if uid does not match real uid or saved uid return EPERM (TODO:
+     * or CAP_SETUID is set, but that was handled in previous clause)*/
+    else if (
+        (uid != thread->uid) && (uid != thread->euid) &&
```

Comment:
The man page doesn't mention thread->euid is to be compared with uid. It says 
> Return EPERM ... if uid does not match the real UID or saved set-user-ID of the calling process.

Reply:
well spotted. thank you.